src/datastore/src/filestore.js

```diff
@@ -232,164 +232,239 @@ export default class FileStore extends NetworkStore {
-        pathname: `${this.pathname}/${metadata._id}`
-      });
-    }
+    return metadata;
+  }
+
+  /**
+   * @private
+   */
+  formUploadResponse(data, file) {
+    delete data._expiresAt;
+    delete data._requiredHeaders;
+    delete data._uploadURL;
+    data._data = file;
+    return data;
+  }
+
+  /**
+   * Upload a file.
+   *
+   * @param {Blob|string} file  File content
+   * @param {Object} [metadata={}] File metadata
+   * @param {Object} [options={}] Options
+   * @return {Promise<File>} A file entity.
+   */
+  upload(file, metadata = {}, options = {}) {
+    metadata = this.transformMetadata(file, metadata);
+    let kinveySaveData = null
 
     // Execute the request
-    return request.execute()
-      .then(response => response.data)
-      .then((data) => {
-        const uploadUrl = data._uploadURL;
-        const headers = new Headers(data._requiredHeaders);
-        headers.set('content-type', metadata.mimeType);
-
-        // Delete fields from the response
-        delete data._expiresAt;
-        delete data._requiredHeaders;
-        delete data._uploadURL;
-
-        // Execute the status check request
-        const statusCheckRequest = new NetworkRequest({
-          method: RequestMethod.PUT,
-          url: uploadUrl,
-          timeout: options.timeout
-        });
-        statusCheckRequest.headers.addAll(headers.toPlainObject());
-        statusCheckRequest.headers.set('Content-Range', `bytes */${metadata.size}`);
-        return statusCheckRequest.execute(true)
-          .then((statusCheckResponse) => {
-            Log.debug('File upload status check response', statusCheckResponse);
-
-            if (statusCheckResponse.isSuccess() === false) {
-              throw statusCheckResponse.error;
-            }
-
-            if (statusCheckResponse.statusCode !== 308) {
-              return file;
-            }
-
-            // Upload the file
-            options.start = getStartIndex(statusCheckResponse.headers.get('range'), metadata.size);
-            return this.uploadToGCS(uploadUrl, headers, file, metadata, options);
-          })
-          .then((file) => {
-            data._data = file;
-            return data;
-          });
+    return this.saveFileMetadata(options, metadata)
+      .then((response) => {
+        kinveySaveData = response.data;
+        return this.makeStatusCheckRequest(response.data, metadata, options.timeout);
+      })
+      .then((response) => {
+        Log.debug('File upload status check response', response);
+
+        if (response.isSuccess() === false) {
+          return Promise.reject(response.error);
+        }
+
+        if (response.statusCode !== 308) {
+          // TODO: Here we should handle redirects according to location header, but this generally shouldn't happen
+          const err = this.getUnexpectedStatusCheckResponsCode(response);
+          return Promise.reject(err);
+        }
+
+        const uploadOptions = {
```

Comment:
I had problems separating which methods care about which options, same with the responses, since objects were being mutated and I wanted to prevent this, so it's clear and no objects get changed by reference since it can get messy. I checked the docs and api reference but didn't see what options can be passed by the user, so I kept the ones which made the most sense - "timeout" and "maxBackoff". Any others I should pass along to the upload? The others seemed internal.